java/ql/src/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -37,10 +37,187 @@ private class ApacheCollectionsModel extends SummaryModelCsv {
   override predicate row(string row) {
     row =
       [
+        // Note that when lambdas are supported we should model things relating to Closure, Factory, Transformer, FluentIterable.forEach, FluentIterable.transform
+        "org.apache.commons.collections4;ArrayStack;true;peek;;;Element of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;ArrayStack;true;pop;;;Element of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;ArrayStack;true;push;;;Argument[0];Element of Argument[-1];value",
+        "org.apache.commons.collections4;Bag;true;add;;;Argument[0];Element of Argument[-1];value",
+        "org.apache.commons.collections4;Bag;true;uniqueSet;;;Element of Argument[-1];Element of ReturnValue;value",
+        "org.apache.commons.collections4;BidiMap;true;getKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;BidiMap;true;removeValue;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.apache.commons.collections4;BidiMap;true;inverseBidiMap;;;MapKey of Argument[-1];MapValue of ReturnValue;value",
+        "org.apache.commons.collections4;BidiMap;true;inverseBidiMap;;;MapValue of Argument[-1];MapKey of ReturnValue;value",
```

Comment:
To check: BidiMap specialises the `values()` method to yield a `Set<V>` (from `Collection<V>` specified in `Map`); do we correctly inherit the model?

Reply:
Yes, we do. I added a model for that method, generated a test for it, removed the model for it and confirmed that the test passes.